sdk/core/Azure.Core/src/SyncAsyncEventArgs.cs

```diff
@@ -0,0 +1,94 @@
+        /// Diagnosing .NET Core ThreadPool Starvation with PerfView</see> for
+        /// a detailed explanation of how that can cause ThreadPool starvation
+        /// and serious performance problems.
+        /// </para>
+        /// <para>
+        /// You can use this <see cref="RunSynchronously"/> property to check
+        /// how the event is being raised and implement your handler
+        /// accordingly.  Here's an example handler that's safe to invoke from
+        /// both sync and async code paths.
+        /// <code snippet="Snippet:Azure_Core_Samples_EventSamples_CombinedHandler">
+        /// var client = new AlarmClient();
+        /// client.Ring += async (SyncAsyncEventArgs e) =&gt;
+        /// {
+        ///     if (e.RunSynchronously)
+        ///     {
+        ///         Console.WriteLine(&quot;Wake up!&quot;);
+        ///     }
+        ///     else
+        ///     {
+        ///         await Console.Out.WriteLineAsync(&quot;Wake up!&quot;);
+        ///     }
+        /// };
+        ///
+        /// client.Snooze(); // sync call that blocks
+        /// await client.SnoozeAsync(); // async call that doesn&apos;t block
+        /// </code>
+        /// </para>
+        /// </remarks>
+        public bool RunSynchronously { get; }
+
+        /// <summary>
+        /// Gets a cancellation token related to the original operation that
+        /// raised the event.  It's important for your handler to pass this
+        /// token along to any asynchronous or long-running synchronous
+        /// operations that take a token so cancellation (via something like
+        /// <code>
+        /// new CancellationTokenSource(TimeSpan.FromSeconds(10)).Token
+        /// </code>
+        /// for example) will correctly propagate.
+        /// </summary>
+        public CancellationToken CancellationToken { get; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SyncAsyncEventArgs"/>
+        /// class.
+        /// </summary>
+        /// <param name="runSynchronously">
+        /// A value indicating whether the event handler was invoked
+        /// synchronously or asynchronously.  Please see
+        /// <see cref="Azure.Core.SyncAsyncEventHandler{T}"/> for more details.
+        /// </param>
+        /// <param name="cancellationToken">
+        /// A cancellation token related to the original operation that raised
+        /// the event.  It's important for your handler to pass this token
+        /// along to any asynchronous or long-running synchronous operations
+        /// that take a token so cancellation will correctly propagate.  The
+        /// default value is <see cref="CancellationToken.None"/>.
+        /// </param>
+        public SyncAsyncEventArgs(bool runSynchronously, CancellationToken cancellationToken = default)
```

Comment:
We need to update our static analyzer to ensure that parameter has correct value

Reply:
Yes, I think so.  I'll file a bug.  Because we won't have many of these, it'll probably be a lower priority.

Reply:
https://github.com/Azure/azure-sdk-tools/issues/1357